### ISSUE TITLE
Realtek Settings Backup

### ISSUE THREAD

**Issue by tita2000** [AlanFinotty/AAFDCHDriverMod#76]
I love this Driver assembly as it supports my devices without any problems and I always install the latest released version. The only thing that is a bit annoying is that I can not backup the room-settings somehow. As I am using a 5.1 speaker system I have to reset the settings every time I updated the driver. It would be great enhancement to have the possibility to backup this data. Or may be there is a way to do this, that I totally missed? Thank you.

**Comment by RCCore** [comment]
> 
> 
> I love this Driver assembly as it supports my devices without any problems and I always install the latest released version. The only thing that is a bit annoying is that I can not backup the room-settings somehow. As I am using a 5.1 speaker system I have to reset the settings every time I updated the driver. It would be great enhancement to have the possibility to backup this data. Or may be there is a way to do this, that I totally missed? Thank you.

Realtek does not offer the backup method for settings.

**Comment by tita2000** [comment]
Alright, but they must be stored somewhere so that one could backup them manually, or by script?

**Comment by RCCore** [comment]
> 
> 
> Alright, but they must be stored somewhere so that one could backup them manually, or by script?

I dont think that is possible, I guess it is stored somewhere but it is not accessible.

**Comment by icedterminal** [comment]
## Preface

I just started digging into this. I had been using a driver version from last year. Long story short I noticed none of the settings are saved. Here's what I have found so far using some debug tools from SysInternals.

If you go to `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render` you will see all the IDs of your audio devices. (I already knew this existed since I had to edit these for VM reasons to make it look real.)

My device key was `{e8aed749-b26a-4c53-bda5-4e02b213b3fa}`. I can verify that by looking under `Properties` key and finding `Analog SPK Output` in the right pane.

When you change the EQ with the GUI, under `FxProperties` key, these entries changed:

- `{4b361010-def7-43a1-a5dc-071d955b62f7},5` = Default EQ is hex `0`. Custom EQ is hex `3ff`. This entry is like an On/Off switch for the EQ.
- `{4b361010-def7-43a1-a5dc-071d955b62f7},6` = The default EQ values vary between `A 1` and `A Z`, but a large amount of values are in fact changing when you set a custom EQ. This is what is actually keeping your EQ retained.

But you can't copy and paste values or else the software bugs out. When looking at what the GUI does, you can see:

`RegSetValue` From `\REGISTRY\A\{59d407b5-1884-462f-d2f3-273cd65cf25f}\LocalState\EQ\[icedterminal]` to `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render\FxProperties\{4b361010-def7-43a1-a5dc-071d955b62f7},6`

![image](https://i.imgur.com/Kxynkvj.png)

Just have to find out where they are being pulled form. `RtkUWP.exe` has to get them from somewhere. They aren't in the registry though. I searched for `{59d407b5-1884-462f-d2f3-273cd65cf25f}` and came up empty. Furthermore, since I had never seen this path before, I gave a look over SysInternals docs to see if there was anything I could find.

## Results

The values for the EQ are being pulled from an application hive. Not the user or system hive, which is what `regedit` loads by default. Applications can use private hives that are loaded on demand. At first, this appears like an issue:

> In contrast to other types of registry hives, application hives are loaded under the \REGISTRY\A registry path name instead of under \REGISTRY\MACHINE or \REGISTRY\USER. The \REGISTRY\A path name is special in that there is no way to traverse this path, and an attempt to open a key under \REGISTRY\A will fail with error status STATUS_ACCESS_DENIED. The only way an application can access a key in an application hive is to use the handle to the root key of the application hive. The application gets this handle from the RegLoadAppKey call that loads the hive.

^ [source](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/filtering-registry-operations-on-application-hives)

The good news is the path the GUI was pulling from was a hint. `LocalState` is directory found in `%LocalAppData%\Packages\<AppName>`. I dug through these folder and found what I was after. The solution is to load the private hive with the registry editor manually.

1. Close Realtek Audio Console and open regedit.
2. Click `HKEY_USERS`.
3. Click [File] > [Load hive...]
4. Realteks private hive is `settings.dat` located at `%LocalAppData%\Packages\RealtekSemiconductorCorp.RealtekAudioControl_dt26b99r8h8gj\Settings`
5. Give it a name. Such as `rtk` or `realtek`.
6. From there you can export your EQ settings to a `.reg` file. Just right click on the hive name and export.
7. UNLOAD THE HIVE! This is important before you close regedit.

![image2](https://i.imgur.com/zO22USk.png)

Uinstall old AAF and install updated AAF. Start Realtek Audio Console once. Close, load the hive, import, unload hive. Your settings will be restored. This is very tedious but the only way to do it. This is the nature of it all.
